File: backend/app/schemas/best_time.py
from pydantic import BaseModel, computed_field

DAY_NAMES = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]


class BestTimeSlot(BaseModel):
    day_of_week: int
    hour_utc: int
    avg_engagement_rate: float
    avg_impressions: float
    sample_count: int

    @computed_field  # type: ignore[prop-decorator]
    @property
    def day_name(self) -> str:
        if 0 <= self.day_of_week <= 6:
            return DAY_NAMES[self.day_of_week]
        return "Unknown"

    model_config = {"from_attributes": True}


class BestTimesResponse(BaseModel):
    account_id: str
    platform: str
    best_times: list[BestTimeSlot]


class HeatmapCell(BaseModel):
    day_of_week: int
    hour_utc: int
    value: float  # engagement rate

    model_config = {"from_attributes": True}


class HeatmapResponse(BaseModel):
    account_id: str
    platform: str
    data: list[HeatmapCell]


class AnalyzeResponse(BaseModel):
    account_id: str
    platform: str
    slots_updated: int
    message: str

File: frontend/src/api/bestTime.ts
import { api } from './client';

export interface BestTimeSlot {
  day_of_week: number;
  hour_utc: number;
  avg_engagement_rate: number;
  avg_impressions: number;
  sample_count: number;
  day_name: string;
}

export interface HeatmapCell {
  day_of_week: number;
  hour_utc: number;
  value: number;
}

export const getBestTimes = (accountId: string) =>
  api.get<{ account_id: string; platform: string; best_times: BestTimeSlot[] }>(
    `/best-times/${accountId}`
  ).then(r => r.data);

export const getHeatmap = (accountId: string) =>
  api.get<{ account_id: string; platform: string; data: HeatmapCell[] }>(
    `/best-times/${accountId}/heatmap`
  ).then(r => r.data);

export const analyzeBestTimes = (accountId: string) =>
  api.post(`/best-times/${accountId}/analyze`).then(r => r.data);
